Lib/glyphsLib/builder/sources.py

```diff
@@ -102,38 +102,45 @@ def _to_designspace_source_layer(self):
                 and "}" in layer.name
                 and ".background" not in layer.name
             ):
-                associated_master = self._sources[layer.associatedMasterId]
-                layer_name_to_master_filenames[layer.name].add(
-                    associated_master.filename
-                )
+                # associated_master = self._sources[layer.associatedMasterId]
+                layer_name_to_master_ids[layer.name].add(layer.associatedMasterId)
                 layer_name_to_glyph_names[layer.name].append(glyph.name)
 
     designspace = self._designspace
-    for layer_name, master_filenames in layer_name_to_master_filenames.items():
-        layer_coordinates = [
+    for layer_name, master_ids in layer_name_to_master_ids.items():
+        # Construct coordinates first, may need to be filled up further down with
+        # defaults.
+        brace_coordinates = [
             int(c)
             for c in layer_name[
                 layer_name.index("{") + 1 : layer_name.index("}")
             ].split(",")
         ]
-        if len(layer_coordinates) > len(designspace.axes):
-            logger.warning(
-                "Glyph(s) %s, brace layer '%s' defines more locations than "
-                "there are design axes.",
-                layer_name_to_glyph_names[layer_name],
-                layer_name,
-            )
 
-        # If we have more locations than axes, ignore the extra locations. When we have
-        # less, the Designspace specification assumes default values for the extra axes.
-        layer_coordinates_mapping = {
-            axis.name: location
-            for axis, location in zip(designspace.axes, layer_coordinates)
-        }
+        for master_id in master_ids:
+            master = self._sources[master_id]
+            master_coordinates = brace_coordinates
+            if len(master_coordinates) < len(designspace.axes):
+                master_locations = [master.location[a.name] for a in designspace.axes]
+                master_coordinates = (
+                    brace_coordinates + master_locations[len(brace_coordinates) :]
```

Comment:
maybe inline that list comprehension in here and get rid of the local variable `master_locations`